app/assets/javascripts/controllers/security_group/security_group_form_controller.js

```diff
@@ -1,54 +1,100 @@
+  var vm = this;
+
+  var init = function() {
+    vm.securityGroupModel = {
+      name: "",
+      description: "",
+      cloud_tenant_name: "",
+      firewall_rules: [],
+    };
+
+    getSecurityGroups();
+    vm.hostProtocols = ["", "TCP", "UDP", "ICMP"];
+    vm.networkProtocols = ["IPV4", "IPV6"];
+    vm.directions = ["inbound", "outbound"];
+
+    vm.formId = securityGroupFormId;
+    vm.model = "securityGroupModel";
+    vm.newRecord = securityGroupFormId == "new";
+    vm.afterGet = false;
+    vm.modelCopy = angular.copy( vm.securityGroupModel );
+    vm.saveable = miqService.saveable;
+    ManageIQ.angular.scope = vm;
+
+    if (vm.newRecord) {
+      vm.securityGroupModel.name = "";
+    } else {
+      miqService.sparkleOn();
+      API.get("/api/security_groups/" + securityGroupFormId + "?attributes=cloud_tenant,firewall_rules").then(function(data) {
+        vm.securityGroupModel.name = data.name;
+        vm.securityGroupModel.description = data.description;
+        vm.securityGroupModel.cloud_tenant_name = data.cloud_tenant.name;
+        vm.securityGroupModel.firewall_rules = data.firewall_rules;
+        vm.afterGet = true;
+        vm.modelCopy = angular.copy( vm.securityGroupModel );
+      }).catch(miqService.handleFailure);
       miqService.sparkleOff();
+    }
+  };
+
+  var getSecurityGroups = function() {
+    API.get("/api/security_groups/?expand=resources&attributes=ems_ref,id,name").then(function(data) {
+      vm.security_groups_list = data.resources;
     }).catch(miqService.handleFailure);
-  }
+  };
 
-  $scope.addClicked = function() {
+  vm.addClicked = function() {
     var url = 'create/new' + '?button=add';
-    miqService.miqAjaxButton(url, $scope.securityGroupModel, { complete: false });
+    miqService.miqAjaxButton(url, vm.securityGroupModel, { complete: false });
+  };
+
+  vm.addFirewallRuleClicked = function() {
+    var index = vm.securityGroupModel.firewall_rules.length;
+    vm.securityGroupModel.firewall_rules[index] = {
+      id: null,
+      resource_id: securityGroupFormId,
+      resource_type: "SecurityGroup",
+      direction: null,
+      ems_ref: null,
+      end_port: null,
+      host_protocol: null,
+      network_protocol: null,
+      port: null,
+      source_ip_range: null,
+      source_security_group_id: null,
+    };
+    vm.angularForm.$setDirty();
   };
 
-  $scope.cancelClicked = function() {
-    if (securityGroupFormId == 'new') {
+  vm.cancelClicked = function() {
+    if (vm.newRecord) {
       var url = '/security_group/create/new' + '?button=cancel';
     } else {
       var url = '/security_group/update/' + securityGroupFormId + '?button=cancel';
     }
     miqService.miqAjaxButton(url);
   };
 
-  $scope.saveClicked = function() {
+  vm.deleteFirewallRuleClicked = function(index) {
+    vm.securityGroupModel.firewall_rules[index].deleted = true;
+    vm.angularForm.$setDirty();
```

Comment:
Can you remove this line?
We don't need to set the form to dirty.

(Besides, `vm.angularForm.$setDirty();` is incorrect and won't work.)